### PR TITLE
WIP: OpenXR integration

### PR DESCRIPTION
**Pull request by blaind**
**Connections**

One take at resolving OpenXR support (#602). See https://github.com/blaind/xrbevy

The code in this pull request is not even close to merge-able, but I thought its good idea to open it for further discussion.

See also https://github.com/gfx-rs/gfx/pull/3761 and https://github.com/gfx-rs/wgpu-rs/pull/910

**Description**
For architecture of the pull request, see https://github.com/blaind/xrbevy/blob/main/docs/architecture.md (evolving document)

**Testing**
(TODO)

### CODE REVIEW

File: wgpu-core/Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,16 @@ trace = ["ron", "serde", "wgt/trace", "naga/serialize"]
 replay = ["serde", "wgt/replay", "naga/deserialize"]
 # Enable serializable compute/render passes, and bundle encoders.
 serial-pass = ["serde", "wgt/serde", "arrayvec/serde"]
+use-openxr = [
+    "openxr",
+    "hal/use-openxr",
+    "wgt/use-openxr",
+    "gfx-backend-vulkan/use-openxr",
```
Comment (blaind) [inline wgpu-core/Cargo.toml:26]: Current PR supports only Vulkan-based OpenXR (are others possible?)

File: wgpu-types/Cargo.toml [hunk 2]
```diff
@@ -14,7 +14,9 @@ license = "MPL-2.0"
 [features]
 trace = ["serde"]
 replay = ["serde"]
+use-openxr = ["openxr"]
 
 [dependencies]
 bitflags = "1.0"
 serde = { version = "1.0", features = ["serde_derive"], optional = true }
+openxr = { version = "0.14", features = ["loaded"], default-features = false, optional = true }
```
Comment (blaind) [inline wgpu-types/Cargo.toml:22]: Eventually, is it okay to have openxr as an optional dependency?

File: wgpu-types/src/openxr.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+impl Default for OpenXROptions {
+    fn default() -> Self {
+        OpenXROptions {}
+    }
+}
+
+/// FIXME FIXME FIXME is this okay?
+pub struct OpenXRHandles {
```
Comment (blaind) [inline wgpu-types/src/openxr.rs:13]: These OpenXR concepts are required at engine-space (bevy) in order to communicate with openxr. This is one take at transferring them between

File: wgpu-types/src/lib.rs [hunk 4]
```diff
@@ -17,6 +17,10 @@
 use serde::{Deserialize, Serialize};
 use std::{num::NonZeroU32, ops::Range};
 
+#[cfg(feature = "use-openxr")]
+/// OpenXR
+pub mod openxr;
```
Comment (blaind) [inline wgpu-types/src/lib.rs:22]: Naming clash with openxr crate, other names?

File: wgpu-core/src/device/mod.rs [hunk 5]
```diff
@@ -2839,6 +2887,16 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         device_id: id::DeviceId,
         desc: &resource::TextureDescriptor,
         id_in: Input<G, id::TextureId>,
+    ) -> (id::TextureId, Option<resource::CreateTextureError>) {
+        self.device_create_texture_raw_image::<B>(device_id, desc, None, id_in)
+    }
+
+    pub fn device_create_texture_raw_image<B: GfxBackend>(
```
Comment (blaind) [inline wgpu-core/src/device/mod.rs:2894]: The logic here is modeled from https://github.com/Ralith/openxrs/blob/5283daa09ddbe7170069ff3ea9e42e8a8a9ff4bc/openxr/examples/vulkan.rs#L537 - basically a raw image is required for the swapchain

File: wgpu-core/Cargo.toml (hunk 1 above)
Comment (agrande) [reply wgpu-core/Cargo.toml:26]: I think they are, at least dx and gl are supported by OpenXR.

Comment (ColdIce1605) [reply wgpu-core/Cargo.toml:26]: from what I've seen from the docs its GL and VK

**Comment by blaind** [comment]
I'll close this - there's a better encapsulation approach with raw handles that's being continued at gfx-rs/wgpu#1609

**Pull request closed without merging**